Reduce potential merge conflicts

eloch216 opened the pull request:
This PR makes some significant changes to the repository. With these changes, the code in the repository will no longer compile, and it will also not define an R package. Instead, there is now an R script (`module_library_setup.R`) that will generate a few missing module-library-specific files for a user. Instructions can be found in `README_skeleton.md`.

There are a few things I am a little unsure about, and I think @gsrohde and @justinmcgrath probably know more about them than me.
- One point is that I made the customization script using R code. This seemed like a decent choice to me, since it should work on any operating system, and presumably any potential users of the skeleton package will know how to run an R script. At minimum, they will have R installed. Nevertheless, maybe there is a better way to implement this script.
- The other point is that I needed to define two C macros: `R_PACKAGE_NAME` and `R_INIT_FUNC`. It seems like it _should_ be possible to generate `R_INIT_FUNC` from `R_PACKAGE_NAME`, but I don't know how to do it. I'm not very familiar with preprocessor directives.
- Is there a reason to have an R-package-style `DESCRIPTION` file for the skeleton repository? It would be nice to have something that clearly specifies the authors, but it doesn't have to be done the same way as in `DESCRIPTION_skeleton`.

While working on this, I realized a few other advantages of having an incomplete R package and a customization script:
- Any forked / customized package derived from this one will include a record of the version of `skelBML` that was used (via `DESCRIPTION_skeleton` and `NEWS_skeleton.md`), as long as the new package maintainer doesn't delete them.
- It should be possible to enable other customizations in the future if they become necessary; we can just modify `module_library_setup.R` and any package maintainers can rerun it.

Let me know what you think.

justinmcgrath commented:
* Script

A script to create the files that contain the package name seems like a good idea. I think that if they're at the point of making one of these packages, they probably have R. If you wanted to avoid scripts, you could provide the templates and the user could change the names manually.

* Macros

I never remember the rules for macro expansion. There's a description here: https://stackoverflow.com/questions/5256313/c-c-macro-string-concatenation. I think this would work.

```c
#define PPCAT_NX(A, B) A ## B
#define PPCAT(A, B) PPCAT_NX(A, B)

#define R_PACKAGE_NAME MyPack
#define R_INIT_FUNC PPCAT(R_init_, R_PACKAGE_NAME)
```

Then "MyPack" is only in one place in that file, but since a script is writing this, what you have also seems fine since there's already a single place that the name gets entered.

I think it would make more sense to use include guards around the entire file as is typically done with header files, rather than having guards around each definition.

* DESCRIPTION file

A file is probably useful, but what's in it could be confusing if people keep it. Maybe it could be written in the past tense, like "This library was created from skelBML ...", but then it's kind of confusing in the skelBML repository.

I can't think of anything for the README. Maybe this would help overall? https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository

justinmcgrath commented:
I don't know whether the include guards in module_library should be package specific. If someone wanted to compile this into a single C++ program, rather than talking through R, it wouldn't work with the new changes.

eloch216 commented:
> I don't know whether the include guards in module_library should be package specific. If someone wanted to compile this into a single C++ program, rather than talking through R, it wouldn't work with the new changes.

I was a bit unsure about this. I didn't think I would be able to define the header guard with macros, so I just set it to something more generic. However, now I see that the startup script could probably generate `module_library.h` with a package-specific header guard since it's a pretty small file. On the other hand...

I think it would be possible for a C++ user to compile the modules as a dynamic library, and then include that library with their other code. I think in that case the header guard wouldn't matter anymore. Even with a package-specific header guard, I think there would be other issues with trying to compile multiple module libraries in one C++ project because each project would want to have its own definition of the `R_PACKAGE_NAME` macro. If the libraries end up using the same value of `R_PACKAGE_NAME` and each have a module with the same base name, the module classes will no longer be properly namespaced.

justinmcgrath commented:
I'm not sure what all it would entail. It could probably be done with macros, and I suspect the R_PACKAGE_NAME macro wouldn't be an issue since each library would include only one package name header file. With dynamic or static linking to compiled libraries, I think there'd need to be header files to know what functions are in the library, and then it's back to conflicts. However, since there are no plans for this, and someone would need to try it to sort out the details, there's no point implementing guesses. If it creates conflicts later, we can rethink things.

gsrohde commented:
I'm still looking at this, but for what it's worth, in case you haven't read this already, here's the GitHub scoop on READMEs: https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-readmes

gsrohde commented:
> Yufeng sent a link earlier that seems to indicate that we can't tell the web version to display `README_skeleton.md`: https://stackoverflow.com/questions/69188881/rename-readme-md-to-custom-name-and-use-as-a-project-description. Maybe there is an order of precendence, so the skeleton repo could have a `README` and then if a `README.md` is added, the "fancier" version would be displayed instead? Not sure if this would be confusing, or if it would even work.

Regarding this specific question:

It is possible to have a _plain_ README in the top-level directory, which will get rendered on GitHub if `README.md` doesn't exist.  The problem with this solution is that any Markdown in the plain README won't be rendered as Markdown.

A better solution might be to use `docs/README.md` for `biocro/skelBML`'s README.  The user browsing to the `biocro/skelML` repository will see this on the repository home page.  (They definitely should see _some_ information on the home page so they know what the repository is about.)  The user fleshing out the skeleton can then override this by supplying a top-level `README.md` file, which will then be given precedence by GitHub.

Another possible solution might be to use a plain README for `biocro/skelBML` which simply tells the user to browse to some Markdown file (`INFO.md`?) which contains the real information.  Again, the user fleshing out the skeleton would supply their own `README.md` file to override this and give information about _their_ package.

eloch216 commented:
> > Yufeng sent a link earlier that seems to indicate that we can't tell the web version to display `README_skeleton.md`: https://stackoverflow.com/questions/69188881/rename-readme-md-to-custom-name-and-use-as-a-project-description. Maybe there is an order of precendence, so the skeleton repo could have a `README` and then if a `README.md` is added, the "fancier" version would be displayed instead? Not sure if this would be confusing, or if it would even work.
> 
> Regarding this specific question:
> 
> It is possible to have a _plain_ README in the top-level directory, which will get rendered on GitHub if `README.md` doesn't exist. The problem with this solution is that any Markdown in the plain README won't be rendered as Markdown.
> 
> A better solution might be to use `docs/README.md` for `biocro/skelBML`'s README. The user browsing to the `biocro/skelML` repository will see this on the repository home page. (They definitely should see _some_ information on the home page so they know what the repository is about.) The user fleshing out the skeleton can then override this by supplying a top-level `README.md` file, which will then be given precedence by GitHub.
> 
> Another possible solution might be to use a plain README for `biocro/skelBML` which simply tells the user to browse to some Markdown file (`INFO.md`?) which contains the real information. Again, the user fleshing out the skeleton would supply their own `README.md` file to override this and give information about _their_ package.

I like the `docs/README.md` option. I think it's nicer than the first option you mentioned because we can still use markdown for it. And it's easier for a user than the third option since the user doesn't need to follow a link. The startup script will automatically produce a top-level `README.md` for the new package that will be displayed instead of the skeleton `README.md`.

eloch216 commented:
> There would be conflicts with the package name file since that's included in module_library.h, and it can't be included in the namespace, since the namespace name comes from the package name. A program would need to include module_library.h from each library, so the package names would conflict. I think the only solution is defining unique names for the include guards in that file.

I modified the startup script so it now uses package-specific header guards in `src/module_library/module_library.h` and `src/R_PACKAGE_NAME.h`. I also changed `R_PACKAGE_NAME.h` to only use one guard around the entire file as you suggested.

eloch216 commented:
I modified `DESCRIPTION_skeleton` so it doesn't look exactly like an R package `DESCRIPTION` file. I also referenced it in the auto-generated `README.md` to indicate that it describes the skeleton module library, not whatever new library was forked from it. While doing this, I also realized that we should probably have a separate license file for the skeleton library in case the new package uses a different license. (Although that would be difficult to do with our current GPL license.)

eloch216 commented:
> > One point is that I made the customization script using R code. This seemed like a decent choice to me, since it should work on any operating system, and presumably any potential users of the skeleton package will know how to run an R script. At minimum, they will have R installed. Nevertheless, maybe there is a better way to implement this script.
> 
> I agree that for portability reasons, R scripts are the way to go. But because I hate having to drop into an R session just to run a script (and because I enjoy writing scripts), I made a revised script that is runnable either using `Rscript` or in R using `source`. I put it on the branch `revised_script` for your perusal. It would have been considerably simpler to have it be "command line only", but I was afraid you'd balk at my _taking away_ the ability to run it using `source`, plus I enjoy a challenge. (Also, I'm not absolutely sure that the code to make the script interactive when run from the command line works in Windows. Perhaps you can test this.)

Thanks, this is very nice. It seems easier to maintain and update than my version, and more robust. I tried running the script from the command line in Windows, and it worked for me without any issues.

I'm glad you didn't take away the ability to run the script with `source`. In the interest of accessibility, I would prefer to make minimal assumptions about what the user is familiar with. We are require some degree of familiarity with R in order to use BioCro, so it's nice to have an option that doesn't require any additional familiarity with running things from the command line outside of R. I think many Windows users (such as myself) don't use the command line very often and would find it more natural to stay inside R.

> Note that this script is based on the version of things before your latest changes. I didn't want to spend any more time on it before getting some feedback.

I can work on updating it to accommodate the more recent changes to this branch.

> Note also that I enforce some rules on package names that you don't.

I think these new rules are helpful. However, they need to be changed a little:
- We should not allow periods in the package name. They are permissible in R package names, but _not_ in R packages that use C code with registered routines. For such packages, there must be a C function called `R_init_NAME`, where `NAME` must exactly match the name of the R package. (See `src/init.cpp`.) If there is a period in `NAME`, this will not be a valid C function name, and the package will not compile. I tested this out and confirmed that periods are a problem.
- The name length limit needs to be 7 characters. If it is any longer, `R CMD check` will complain about path lengths that exceed the limit for tarballs. This is due to the presence of the boost library. I think it is also helpful to explain why this limit is in place, rather than just saying "overly long names are not recommended."

> The script also removes all of the complicated cobbling together of strings in favor of using template files, where the user-supplied package name is simply plugged in at the appropriate places. This seems much clearer and much more maintainable to me.

This is great!

> And just to prove that it can be done, I use C macros to define the init function from the package name. But I strongly recommend avoiding the use of C macros (except for header guards), and simply revising the script to write _all_ files that depend on the package name. This recommendation I base on what Bjarne Stroustrup (the creator of C++) has to say about them:
> 
> > Macros are very important in C but have far fewer uses in C++. The first rule about macros is: don’t use them unless you have to. Almost every macro demonstrates a flaw in the programming language, in the program, or in the programmer. Because they rearrange the program text before the compiler proper sees it, macros are also a major problem for many programming support tools. So when you use macros, you should expect inferior service from tools such as debuggers, cross-reference tools, and profilers. If you must use macros, please read the reference manual for your own implementation of the C++ preprocessor carefully and try not to be too clever. Also, to warn readers, follow the convention to name macros using lots of capital letters.
> 
> [Stroustrup, Bjarne. The C++ Programming Language . Pearson Education. Kindle Edition. ]

The original reason to use macros was to reduce the number of times the user needed to type the package name. But, as you point out, the startup script makes this less important, since the user only needs to type the name once regardless of how many times it is used throughout the package source code.

There may be other benefits of using macros though. For example, suppose we make a change to `src/R_module_library.cpp` in the skeleton module library repository, perhaps to the `R_get_all_modules` function's internal code. How will maintainers of other module library packages update their repositories to include the changes? If `src/R_module_library.cpp` uses a macro to specify the library (via `using library = R_PACKAGE_NAME::module_library;`), then maintainers just need to use git to update `R_module_library.cpp` to the latest version. If instead `R_module_library.cpp` is generated from a template via the script and has a hard-coded package name (like `using library = myLib::module_library;`), then maintainers will need to use git to get the latest version of the template and then re-run the startup script to update their repositories.

Maybe the extra step when updating isn't a big deal, and it might be better to avoid the macros. It's good to consider all the angles, though.

justinmcgrath commented:
> I agree that for portability reasons, R scripts are the way to go. But because I hate having to drop into an R session just to run a script (and because I enjoy writing scripts), I made a revised script that is runnable either using `Rscript` or in R using `source`. I put it on the branch `revised_script` for your perusal. It would have been considerably simpler to have it be "command line only", but I was afraid you'd balk at my _taking away_ the ability to run it using `source`, plus I enjoy a challenge. (Also, I'm not absolutely sure that the code to make the script interactive when run from the command line works in Windows. Perhaps you can test this.)

This is nice. I have only the smallest of comments, which is that I like to front-load branches, so I might have done this instead somewhere at the top level.

```r
smartReadLines <- if (interactive()) {
    readline
} else {
    function() {
        input <- file("stdin")
        answer <- readLines(input, 1) 
        on.exit(close(input))
        return(answer)
    }
}
```

eloch216 commented:
> I like to front-load branches

What does this mean?

eloch216 commented:
It does seem weird to have a CSV file with no commas, so I will add them

![image](https://user-images.githubusercontent.com/48919455/203378771-5f0f6b61-2fe6-4580-8c8a-074bc6e6c976.png)

justinmcgrath commented:
> > I like to front-load branches
> 
> What does this mean?

Whenever there's a logical branch, like `if` or `switch` statements, I like to resolve them as early as possible in execution. I find that it's simpler to follow. It also makes execution simpler, since then the branch is encountered fewer times, maybe only once. Making separate modules for C3 and C4 and specifying them, rather than something like `photosynthesis(type='c3')` is the same sort of thing. For this short script it doesn't matter, but I noticed it so added my two cents.

gsrohde commented:
> It does seem weird to have a CSV file with no commas, so I will change the extension
> 
> ![image](https://user-images.githubusercontent.com/48919455/203378771-5f0f6b61-2fe6-4580-8c8a-074bc6e6c976.png)

Or you could just rename the file `template_table.table`.

gsrohde commented:
> * The name length limit needs to be 7 characters. If it is any longer, `R CMD check` will complain about path lengths that exceed the limit for tarballs. This is due to the presence of the boost library. I think it is also helpful to explain why this limit is in place, rather than just saying "overly long names are not recommended."

I didn't understand that the 7-character limit was due to the problem with long path names and couldn't understand why you were being so restrictive when many CRAN packages have considerably longer names.

It sure would be nice not to have to include the Boost library.  I don't remember the reasons given for not just using `BH`.  Maybe it doesn't include all of portions of the library that we need?

gsrohde commented:
> I think these new rules are helpful. However, they need to be changed a little

Since the restriction on periods and the length limit are hard-and-fast rules, that will simplify things—we can get rid of the `run_addl_checks` function, which I didn't like much anyway.

eloch216 commented:
I added a version of Scott's script to this branch, where I updated it to include the recent changes I made here. It still uses C macros and the `run_addl_checks` function, though, so it's probably not final.

justinmcgrath commented:
> > * The name length limit needs to be 7 characters. If it is any longer, `R CMD check` will complain about path lengths that exceed the limit for tarballs. This is due to the presence of the boost library. I think it is also helpful to explain why this limit is in place, rather than just saying "overly long names are not recommended."
> 
> I didn't understand that the 7-character limit was due to the problem with long path names and couldn't understand why you were being so restrictive when many CRAN packages have considerably longer names.
> 
> It sure would be nice not to have to include the Boost library. I don't remember the reasons given for not just using `BH`. Maybe it doesn't include all of portions of the library that we need?

If BH is used instead of Boost, then compiling with SimRoot won't work. Would it help to change 'boost_1_71_0' to something shorter, or are only the paths under that the problem? We could call it 'include' and include the version number in a file in that directory.

gsrohde commented:
> > > * The name length limit needs to be 7 characters. If it is any longer, `R CMD check` will complain about path lengths that exceed the limit for tarballs. This is due to the presence of the boost library. I think it is also helpful to explain why this limit is in place, rather than just saying "overly long names are not recommended."
> > 
> > 
> > I didn't understand that the 7-character limit was due to the problem with long path names and couldn't understand why you were being so restrictive when many CRAN packages have considerably longer names.
> > It sure would be nice not to have to include the Boost library. I don't remember the reasons given for not just using `BH`. Maybe it doesn't include all of portions of the library that we need?
> 
> If BH is used instead of Boost, then compiling with SimRoot won't work.

What is SimRoot?

> Boost also doesn't guarantee backwards compatibility, so updates could break BioCro. 

Is this more true of Boost than, say, R itself, or for that matter, C++ and the other C++ libraries we use?

> Would it help to change 'boost_1_71_0' to something shorter, or are only the paths under that the problem? We could call it 'include' and include the version number in a file in that directory.
> 
> Possibly the 'boost' directory could be at the top level. I don't remember the details so maybe that's not allowed though.

I'm actually more concerned about needlessly requiring users to have multiple copies of Boost if, say, they are working on multiple BioCro libraries at once.

justinmcgrath commented:
> What is SimRoot?

It's called OpenSimRoot now, and it's a root model that has been coupled with BioCro: https://gitlab.com/rootmodels/OpenSimRoot. Progress is stilted with this project, but people are still pursuing it. R isn't and cannot be used. Thus, BioCro is treated like a library and compiled directly with OpenSimRoot.
> > Boost also doesn't guarantee backwards compatibility, so updates could break BioCro.
> 
> Is this more true of Boost than, say, R itself, or for that matter, C++ and the other C++ libraries we use?

I think backwards compatibility is an unattainable Valhalla, but lots of projects strive for it. It looks like the R authors try to do it, but I haven't seen an official policy. The Boost authors explicitly state that they do not try to maintain backwards compatibility though.

> I'm actually more concerned about needlessly requiring users to have multiple copies of Boost if, say, they are working on multiple BioCro libraries at once.

I see libraries as a way to ease code reuse, and making things smaller can be a benefit, but is often not important. BioCro has about 50 MB of Boost files, which is certainly not small, but compared to most software today and the size of hard drives, it's hardly worth considering. If we ever take out Boost Units, it will be a bit smaller. If we want to use one copy, then every BioCro developer would need to do a site-wide installation of Boost, which is tedious. Maintenance and installation are both much easier if the necessary files are packaged with BioCro, and the cost of some hard drive space is very low.

Tangentially, since backwards compatibility is impossible, multiple copies of libraries are the way this has been handled. Windows SXS is an example. I also found this project yesterday that keeps multiple versions of R packages: https://groundhogr.com/. It seems like a good idea, but I tried it and it needs some work.

gsrohde commented:
> > What is SimRoot?
> 
> It's called OpenSimRoot now, and it's a root model that has been coupled with BioCro: https://gitlab.com/rootmodels/OpenSimRoot. Progress is stilted with this project, but people are still pursuing it. R isn't and cannot be used. Thus, BioCro is treated like a library and compiled directly with OpenSimRoot.

Do you mean OpenSimRoot is compiled  with the `BioCro.so` or `BioCro.dll` library file, or do you mean "library" in a more general sense—a bunch of source files which must be compiled with the OpenSimRoot source files (similar to the way one must use the Boost library)?  I'm guessing the latter, since with the former, Boost is already included, so it's not an issue.  I guess I don't really understand the issue here.
> 
> > > Boost also doesn't guarantee backwards compatibility, so updates could break BioCro.
> > 
> > 
> > Is this more true of Boost than, say, R itself, or for that matter, C++ and the other C++ libraries we use?
> 
> I think backwards compatibility is an unattainable Valhalla,

[and considering what happened to Valhalla, one probably wouldn't want to attain it in any case :).]

> but lots of projects strive for it. It looks like the R authors try to do it, but I haven't seen an official policy. The Boost authors explicitly state that they do not try to maintain backwards compatibility though.
> 

What they actually say, if we're reading from the same place, is "backward compatibility with prior version [sic] isn't always possible," something which is, if software vendors were being honest, generally the case.  "Isn't always possible" at least suggests that they do try for it when it is possible.

Still, they do take pains to say "Deal with this by freezing the version of the Boost libraries used by your project. Only upgrade at points in your project's life cycle where a bit of change will not cause problems", which suggests maybe you are right to be cautious.

Since the maintainer of `BH` is associated with the _tinyverse_ (as opposed to _tidyverse_) initiative, you would think he would be particularly aware of and sensitive to dependency and versioning concerns.  Perhaps I'll get in touch with him—perhaps he has a solution to this problem that I haven't considered.

(The only solution that occurred to me was forking a release of `BH` corresponding to a suitable release of Boost, then renaming the package to include the version number (so if users already have a copy of `BH`, there wouldn't be a conflict), and making it publicly available.  It seems unlikely we could put such a package on CRAN, though, and if you complained that this solution was a tad messy, I'd probably have to agree.)

> > I'm actually more concerned about needlessly requiring users to have multiple copies of Boost if, say, they are working on multiple BioCro libraries at once.
> 
> I see libraries as a way to ease code reuse, and making things smaller can be a benefit, but is often not important. BioCro has about 50 MB of Boost files, which is certainly not small, but compared to most software today and the size of hard drives, it's hardly worth considering. If we ever take out Boost Units, it will be a bit smaller. If we want to use one copy, then every BioCro developer would need to do a site-wide installation of Boost, which is tedious.

Are you speaking from the perspective of a developer who is not using R?  In R, all one has to do is `install.packages('BH')` , and it just works (assuming that after removing Boost from BioCro source we add `BH` to the DESCRIPTION file's `LinkingTo` field).

eloch216 commented:
I will probably have something to say about this boost discussion later, but for now, can we pause this discussion or move it elsewhere? Let's not allow it to derail this PR, since 99% of it is independent of how the boost library is accessed.

gsrohde commented:
> I will probably have something to say about this boost discussion later, but for now, can we pause this discussion or move it elsewhere? Let's not allow it to derail this PR, since 99% of it is independent of how the boost library is accessed.

I've made an issue for it.

gsrohde commented:
> I modified the startup script so it now uses package-specific header guards in `src/module_library/module_library.h` and `src/R_PACKAGE_NAME.h`.

Have you checked that this actually does what you want it to?

I seem to remember there being problems with compiling multiple libraries together, and using unique header guards didn't solve the problem.  But I may be misremembering, or maybe this was only a problem with a previous version of the code and it has changed in such a way that this is no longer an issue.

In any case, I recommend _not_ using package-specific header guards unless they actually serve the purpose you intend, even though they may otherwise be harmless.  Going against the `filename.h` --> `FILENAME_H` convention is confusing and misleading unless it is actually serving some useful purpose.

gsrohde commented:
> I also found this project yesterday that keeps multiple versions of R packages: https://groundhogr.com/. It seems like a good idea, but I tried it and it needs some work.

My impression is that `renv` is more mature and thus likely more reliable.  But as @eloch216 suggested, further discussion of this can go elsewhere.

eloch216 commented:
> > I modified the startup script so it now uses package-specific header guards in `src/module_library/module_library.h` and `src/R_PACKAGE_NAME.h`.
> 
> Have you checked that this actually does what you want it to?
> 
> I seem to remember there being problems with compiling multiple libraries together, and using unique header guards didn't solve the problem. But I may be misremembering, or maybe this was only a problem with a previous version of the code and it has changed in such a way that this is no longer an issue.
> 
> In any case, I recommend _not_ using package-specific header guards unless they actually serve the purpose you intend, even though they may otherwise be harmless. Going against the `filename.h` --> `FILENAME_H` convention is confusing and misleading unless it is actually serving some useful purpose.

I think it's not so unusual to include the name of a namespace in a header guard to better ensure that it is unique to that file: https://softwareengineering.stackexchange.com/questions/330987/namespaces-and-header-guards-with-naming-conflicts

gsrohde commented:
> I think it's not so unusual to include the name of a namespace in a header guard to better ensure that it is unique to that file: https://softwareengineering.stackexchange.com/questions/330987/namespaces-and-header-guards-with-naming-conflicts

Remind me—does one need to use this header file in order to use a corresponding BioCro library's .dll or .so file?  Is the the _only_ header file needed (per library)?  And is this only needed when one is working in C++ directly rather than via the R interface?  Did we get this working smoothly?

eloch216 commented:
> Remind me—does one need to use this header file in order to use a corresponding BioCro library's .dll or .so file? Is the the only header file needed (per library)? And is this only needed when one is working in C++ directly rather than via the R interface? Did we get this working smoothly?

Here is my understanding of the situation:
- For an R package that defines a single module library, the considerations are pretty simple. In that case, it's just necessary to have header guards that are unique within the `src` directory. It's not even necessary to have the modules and library inside a namespace. Then an R user can access multiple module libraries via the R package for each library. I think this problem is sufficiently solved.
- These questions about header guards and namespaces are relevant for a C++ user who wants to use modules from multiple libraries. In this case I think there are two strategies for doing it.
  - One is to separately compile each library into its own `.dll` or `.so` file, and link to those in a C++ project. In this case, I think each module library (and its associated module classes) needs to be in its own namespace. But header guards probably only need to be unique within each module library.
  - The other is to include all the files from each module library and compile it once as a single C++ project. In this case, I think each module library must be namespaced, and the header guards need to be unique across the project as a whole.
  - In both of these scenarios, the R-specific functions in `R_module_library.h`, `R_module_library.cpp`, and `init.cpp` will not be used for anything.

We would like for all of these scenarios to be possible, so I was thinking it would be good to have unique header guards for files in `src/module_library` that include the namespace name.

eloch216 commented:
Following Scott's earlier comments, I changed the script again to generate files that no longer use C macros or the R package name string. Instead, these things are now hard-coded. I also used header guards that include the namespace in a few places. One possible downside is that `module_library.cpp` is now created by the script from a template, so if a module library repository maintainer needs to rerun the script following changes to the skeleton library, their module table will get overwritten.

gsrohde commented:
> I'm not sure about putting `boost` at the top level. I have a hazy recollection of trying it and running into some problem, but I don't remember what it was.

For what it's worth—and I probably should have done so sooner—I tried this out, and I didn't have any problem.  I tried it in two different ways, and they both worked.

The first was simply to change the `PKG_CPPFLAGS` line in the `Makevars` files to read
```
PKG_CPPFLAGS+=-I.. -DR_NO_REMAP
```
after renaming `inc/boost` to `boost`.  I was somewhat dubious about making the top-level BioCro directory an include directory, but it didn't really seem to be a problem.

The second way was to change the `PKG_CPPFLAGS` line to read
```
PKG_CPPFLAGS+=-I../boost -DR_NO_REMAP
```
and then change all the includes of the form
```
#include <boost/[subdirs]/[boostfile].hpp>
```
to
```
#include <[subdirs]/[boostfile].hpp>
```
At first this seemed like a better solution, but then I decided that there is some documentary value in having "boost" in the `#include` path, and for this reason I don't really like this solution.

I'm not advocating changing anything, just showing that we could.

justinmcgrath commented:
> * One is to separately compile each library into its own `.dll` or `.so` file, and link to those in a C++ project. In this case, I think each module library (and its associated module classes) needs to be in its own namespace. But header guards probably only need to be unique within each module library.
> * The other is to include all the files from each module library and compile it once as a single C++ project. In this case, I think each module library must be namespaced, and the header guards need to be unique across the project as a whole.

The conclusion is the same, but I thought I'd point out that even is a library is dynamically linked, you include a header file for it, so the header guards need to be unique in both of these  cases.

justinmcgrath commented:
> For what it's worth—and I probably should have done so sooner—I tried this out, and I didn't have any problem. I tried it in two different ways, and they both worked.

I agree that not having "boost" in the include path doesn't seem great. I'm also not sure what I think about including the entire directory. It's not an ideal way to organize things. There's probably a compile-time cost, but I think it's negligible. It does seem that the benefit of handling longer path names outweighs the odd organization.

justinmcgrath commented:
Maybe this is too complicated, but there is probably a way to avoid overwriting the user modified module_library.cpp. All of the include directives would need to be in a separate file, and the table of modules would be in a separate file. Then module_library.cpp would look something like this:

```c++
#include "module_library.h"
#include "../framework/module_creator.h"  // for create_mc

// Include all the header files that define the modules.
#include "user_defined_list_of_module_headers.h"

creator_map %1$s::module_library::library_entries =
{
    #include "user_defined_module_table.h"
};
```

I don't love this, but maybe it sparks some better ideas from someone else.

justinmcgrath commented:
Maybe the parts that are generated could be surrounded with some identifying comments. Then the script would only modify those places. Instead of replacing %1$s in the namespace, it would write the entire line.

```c++
// HEADER_START. Do not modify this comment. It's needed to automatically propagate updates.
// Do not modify anything between HEADER_START and HEADER_END, as changes will be removed when updating.
#include "module_library.h"
#include "../framework/module_creator.h"  // for create_mc
// HEADER_END. Do not modify this comment. It's needed to automatically propagate updates.

// Include all the header files that define the modules.
#include "example_module.h"

// NAMESPACE_START. Do not modify this comment. It's needed to automatically propagate updates.
// Do not modify anything between NAMESPACE_START and NAMESPACE_END, as changes will be removed when updating.
creator_map package_name::module_library::library_entries =
// NAMESPACE_END. Do not modify this comment. It's needed to automatically propagate updates.
{
    {"example_module", &create_mc<example_module>}
};
```

This seems a little better than separate files, but I still don't love it.

gsrohde commented:
> Maybe this is too complicated, but there is probably a way to avoid overwriting the user modified module_library.cpp.

As of now, the script backs up any existing version of the file if it differs from what would be written by the script.  I think this is good enough for now—I don't think it imposes too big of a burden on the user to require merging in anything from the backup they may want to keep.  If you come up with an alternative solution you really _do_ love, then maybe we can revisit this.

gsrohde commented:
> > * One is to separately compile each library into its own `.dll` or `.so` file, and link to those in a C++ project. In this case, I think each module library (and its associated module classes) needs to be in its own namespace. But header guards probably only need to be unique within each module library.
> > * The other is to include all the files from each module library and compile it once as a single C++ project. In this case, I think each module library must be namespaced, and the header guards need to be unique across the project as a whole.
> 
> The conclusion is the same, but I thought I'd point out that even is a library is dynamically linked, you include a header file for it, so the header guards need to be unique in both of these cases.

As I hint in my revised version of the "large-scale modifications" guidelines, I think it would be worthwhile writing up either a well-defined API for those who wish to use BioCro at the C++ level (and in particular, for those wishing to use it together with user-defined libraries); or a short "how-to" manual on how to do this.  Or (preferably, I think) both.  Not only do I think this would help clarify and possibly bring to light any outstanding questions and issues, I also think these are things we will want to eventually provide to potential users.
